pretyping/globEnv.ml

```diff
@@ -0,0 +1,198 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Util
+open Pp
+open CErrors
+open Names
+open Environ
+open EConstr
+open Evarutil
+open Termops
+open Vars
+open Ltac_pretype
+
+(** This files provides a level of abstraction for the kind of
+    environment used for type inference (so-called pretyping); in
+    particular:
+    - it supports that term variables can be interpreted as Ltac
+      variables pointing to the effective expected name
+    - it incrementally and lazily computes the renaming of rel
+      variables used to build purely-named evar contexts
+*)
+
+type t = {
+  static_env : env;
+  (** For locating indices *)
+  renamed_env : env;
+  (** For name management *)
```

Comment:
Rather than a full-fledged environment, I think that a rel_context should be enough.

Reply:
This is what I had at some time and I changed it for a full env by symmetry. Anyway, in the last version of this PR used in #7300, I only have one `env` and the renaming data is moved to the `extra` flag. I did not update yet this PR because it is always quite tricky to split a PR into subPRs and to maintain them in a consistent state, but I shall do it at some time when the discussion on the contents is complete.

Reply:
OK, no problem then.